### ISSUE TITLE
Placing Icons on the Bookmarks menu and toolbar

### ISSUE THREAD

**Issue by jefferysmall** [muffinresearch/favicon-customizer#3]
Hi:

Just started using the add-on and have got it working for icons on the tabs which is great.  What I really need to do is get these same icons displayed for these links on the bookmarks toolbar and in the bookmarks menu.  Before the upgrade to Quantum I had an add-on that worked great, but was not migrated forward.  Can this tool be modified to display the icons in these locations as well?

Thanks.
--
Jeffery Small

**Comment by muffinresearch** [comment]
Hi @jefferysmall. The use-case for updating the bookmarks toolbar and menus wasn't something I'd originally considered to be part of this but I can see why it's something that would be useful.

It would need looking into to see if it's possible. I personally don't need that functionality so it would be likely to be low priority from my perspective. That said, if someone wanted to propose a patch or do some investigation as to whether it's viable I'd definitely consider getting that feature added.

**Comment by Smylers** [comment]
I'm also missing having custom icons on bookmarks: it's harder to find the one I'm looking for since the icons disappeared.

Fiddling with the chrome dev tools (`Alt`+`Ctrl`+`Shift`+`I`), I found the bookmarks toolbar has elements like this for menu items:

    <menuitem class="menuitem-iconic bookmark-item menuitem-with-favicon"
      scheme="http" label="Radio 4" image="page-icon:http://www.bbc.co.uk/radio4"/>

and like this for buttons:

    <toolbarbutton class="bookmark-item" label="Weather" scheme="javascript"
      image="page-icon:javascript:var%20page%20=%20'http://www.bbc.co.uk/weather/2646272';%20if%20(new%20Date().getHours()%20&gt;=%2015)%20page%20=%20page%20+%20'/tomorrow';%20location%20=%20page;"
      style="visibility: visible;"/>

Manually changing those `image=` attributes to URLs of the desired icons (`file://` URLs of pngs stored locally in my case) does indeed change the icons that appear in the bookmarks toolbar.

I know `userChrome.css` can be used to tweak CSS properties of those elements (they'd presumably have to be selected by label, something like `.bookmark-item[label="Weather"]`), but that's insufficient for changing an attribute. And my Firefox interface knowledge runs out at this point.

Bookmarks are stored in `places.sqlite`. I don't know if an extension has access to that, but I couldn't work out where the icons are specified anyway. The table `moz_places` has a `favicon_id` field ... but it doesn't seem to refer to anything: [the docs](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Places/Database) mention a `moz_favicons` table, but I don't seem to have that.

There is a separate `favicons.sqlite` database. That has a table of pages, a table of icons, and one linking IDs of them together. The IDs don't correspond to the `favicon_id` field from the other database. The `moz_icons` table has an `icon_url` field and also a `data` blob, which appears to be the actual bytes of the png. I haven't tried changing entries in this database to see if it affects which icons Firefox uses for sites.

Is any of that of use? Is there anything else specific we could help you by researching? Thanks.

**Comment by jefferysmall** [comment]
Smylers, good investigative work.  I'm willing to help test anything that can be done to implement this.

**Comment by jefferysmall** [comment]
ElhemEnohpi

How are you identifying the "label" of the bookmark item?  Is the "label" the same as the "Name" property?  On my toolbar, the Name property is the null string for all buttons so that only the icon displays.  Do I need to reintroduce Names to all these bookmark buttons, use that as the "label" string, and then hide the Names with your final css entry, or am I missing something and the labels come from somewhere else?  Thanks for continuing to look into this.

**Comment by ElhemEnohpi** [comment]
@jefferysmall yes, that's right. The label is the Name. You have to give them names so you can select them, then hide the names with css.

**Comment by jefferysmall** [comment]
Following up to my own post:

I found a solution for question #2 above.  I assumed that the margin settings from my toolbarbutton.bookmark-item definition would be inherited by the custom icon definitions, but that was not the case.  I modified the entries as follows:

.bookmark-item[label="Cal"] image {
    margin: 0px !important;
    padding: 0 0 16px 16px !important;
    background:url("icons/cal_01.png") !important;
}

and the excess space was eliminated.  So that pretty much resolves my problems.  You can ignore the question altogether, or if you have some insight regarding question #1, you can comment and teach me something else new!  Thanks again for the help.

**Comment by xlovinglyx** [comment]
Same here .. . 

If changing favicons isn't accessible/added right clicking on them,
The approach of going to addons, checking properties & rename all my
bookmarks there to choose icons is everything people would run from .. .

We need individual bookmarks right click access for changing each bookmark.
